docs/specs/priority-queue.md

```diff
@@ -1,71 +1,73 @@
-# Priority queue proposal
+﻿# Priority queue proposal
 
 ## User data
 
 Before anything is proposed, let's start with what can be prioritized...
 
 ### Input types
 
-We want to be able to support all of the following configurations:
+There are two possible scenarios in which to use a priority queue:
 
-1. User data *is separate* from the priority (two physical instances).
-2. User data *contains* the priority (as one or more properties).
-3. User data *is* the priority (implements `IComparable<T>`).
-4. Rare case: priority is obtainable via some other logic (resides in an object
-different from the user data).
+1. Each item of user data is *assigned* a *separate* priority (priority is a separate object from the elements in the collection.) We will sometimes call this the *extrinsic priority* scenario.
+2. User data either *is itself a priority*, or has some other *natural* priority which can be derived from the element. (It might have a priority property, a it might be a *computed* priority using some function mapping elements to priorities.) We will sometimes call this the *intrinsic priority* scenario.
 
-In this entire document, I will refer to the cases above with **(1)**, **(2)**,
-**(3)**, and **(4)**.
+### Uniqueness and changes in priority.
 
-### Ideal solution
+This implementation of priority queue is actually designed to provide a DICTIONARY-like or SET-like interface, where each element is unique (like a key), and priorities are mutable. This allows you to easily use the data structure for many popular scenarios requiring priority queues. 
+All elements are equality comparable, and each distinct element (according to equality comparison) can have at most a unique occurence in the queue, and be assigned a single priority value.
+That priority value may be updated, over time.
 
-Obviously, our solution should be flexible enough to (respectively):
+This means there are really two types of comparators (comparison operators) involved in the definition of a priority queue class - one for comparing *elements* pairwise for equality, and one for comparing *element priorities* pairwise to see which is higher priority. It is unnecessary, and usually undesirable for the same comparator to be used for both priority comparisons and equality comparisons.
```

Comment:
I don't really understand why we need 2 different comparers if one comparer should be enough to define if two elements are equal and if an element is greater than the other one based on its priority.

Reply:
@safern there's two completely different types of comparison here. 
Comparison type #1: is this element the same element which was already inserted in the queue? (Used for finding an element to update its priority)
Comparison type #2: is this element a higher priority than that element? (Used for sorting elements so they get dequeued in priority order)

Reply:
@safern This is also why the type parameters of the comparers are different. (One is a comparer of of `<TElement>` the other is of `<TPriority>`.)

Reply:
Wait, nevermind. That's a different thing.

We need the second comparer for both cases - for the second case it is because just because two items have the same priority doesn't mean they're necessarily equal.

Reply:
I see. That way makes sense.

Reply:
> We need the second comparer for both cases - for the second case it is because just because two items have the same priority doesn't mean they're necessarily equal.

That makes sense. I was not taking that in consideration either.

Reply:
> I was not taking that in consideration either.

+1. We should add a comment or reword it so it's a bit more clear that the reason for two comparers is that two elements can have the same priority but be unequal.

Reply:
> We should add a comment or reword it so it's a bit more clear that the reason for two comparers is that two elements can have the same priority but be unequal.

Agree.

Reply:
Revised a couple times, hopefully its clear now.